Applications/CausalLM/layers/mha_core.cpp

```diff
@@ -0,0 +1,767 @@
+        float *input_addr = in.getData<float>() + num_head * head_dim * i;
+        uint16_t *cache_addr = cache.getData<uint16_t>();
+        int row_to_compute = i + 1;
+        size_t out_start_row = (i + 1) * i / 2;
+
+        float *output_addr = out.getData<float>() + out_start_row * num_head;
+
+        futures.emplace_back(pool.submit_task([=]() {
+          nntrainer::compute_kcaches<uint16_t>(
+            input_addr, cache_addr, output_addr, row_to_compute,
+            num_head / group_size, head_dim, group_size, 16);
+        }));
+      }
+      for (auto &fut : futures)
+        fut.get();
+    }
+  } else if (in.getDataType() == ml::train::TensorDim::DataType::FP16) {
+#ifdef ENABLE_FP16
+    if (from) {
+      nntrainer::compute_kcaches(
+        in.getData<_FP16>(), cache.getData<_FP16>(), out.getData<_FP16>(),
+        from + 1, num_head / group_size, head_dim, group_size, 16);
+    } else {
+      std::vector<std::future<void>> futures;
+      for (unsigned int i = 0; i < sequence_len; ++i) {
+        _FP16 *input_addr = in.getData<_FP16>() + num_head * head_dim * i;
+        _FP16 *cache_addr = cache.getData<_FP16>();
+        int row_to_compute = i + 1;
+        size_t out_start_row = (i + 1) * i / 2;
+
+        _FP16 *output_addr = out.getData<_FP16>() + out_start_row * num_head;
+
+        futures.emplace_back(pool.submit_task([=]() {
+          nntrainer::compute_kcaches(input_addr, cache_addr, output_addr,
+                                     row_to_compute, num_head / group_size,
+                                     head_dim, group_size, 16);
+        }));
+      }
+      for (auto &fut : futures)
+        fut.get();
+    }
+#else
+    NNTR_THROW_IF(true, std::invalid_argument) << "enable-fp16 is not set!";
+#endif
+  }
+}
+
+void MHACoreLayer::one_batch_incremental_forwarding(
+  const unsigned int batch, const unsigned int _from, const unsigned int from,
+  const unsigned int to, nntrainer::Tensor &query, nntrainer::Tensor &key,
+  nntrainer::Tensor &value, nntrainer::Tensor &output,
+  nntrainer::Tensor &cache_key, nntrainer::Tensor &cache_value,
+  ml::train::TensorDim &query_dim, ml::train::TensorDim &query_step_dim,
+  ml::train::TensorDim &key_dim, ml::train::TensorDim &value_dim,
+
+  ml::train::TensorDim &cache_key_dim, ml::train::TensorDim &cache_key_step_dim,
+  ml::train::TensorDim &cache_value_dim,
+  ml::train::TensorDim &cache_value_step_dim, ml::train::TensorDim &output_dim,
+  ml::train::TensorDim &output_step_dim) {
+
+  /**
+   *  cache_key
+   *  +--------+                        ->
+   *  |        |                        ->
+   *  |        |                        ->
+   *  |........| from                   ->
+   *  |........| to -> b_cache_key_step -> b_cached_key
+   *  |        |
+   *  +--------+
+   *
+   */
+
+  /** 1. Load Input Tensors of this batch : b_ denotes a Tensor for this batch
+   * **/
+  auto &pool = nntrainer::ThreadPoolManager::getInstance();
```

Comment:
The method is chaned in main.
Please rebase and update this part.
```suggestion
  auto &pool = nntrainer::ThreadPoolManager::Global().getThreadPool();
```

Reply:
Done.